plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java

```diff
@@ -28,6 +47,88 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
 	
 	@Override
 	protected void registerOtherEditorHandlers() {
+		setupOpenType();
+	}
+	
+	// TODO: Evaluate source structure in background thread.
+	private void setupOpenType() {
+		evaluateGlobalSourceStructure();
+		listenToUpdatesOfGlobalSourceStructure();
+	}
+	
+	private void evaluateGlobalSourceStructure() {
+		try {
+			ResourcesPlugin.getWorkspace().getRoot().accept(this::resourceTouched);
+		} catch(Exception e) {
+			AbstractLangCore.log().logError("Global source structure could not be determined", e);
+		}
+	}
+	
+	private boolean resourceTouched(IResource resource) {
+		convertToRustFileLocation(resource).ifPresent(RustEditorActionContributor::fileTouched);
+		return true;
+	}
+	
+	private void listenToUpdatesOfGlobalSourceStructure() {
+		ResourcesPlugin.getWorkspace().addResourceChangeListener(this::processEvent, IResourceChangeEvent.POST_CHANGE);
 	}
 	
+	private void processEvent(IResourceChangeEvent event) {
+		try {
+			event.getDelta().accept(this::applyResourceDelta);
+		} catch(Exception e) {
+			AbstractLangCore.log().logError("Global source structure could not be determined", e);
+		}
+	}
+	
+	private boolean applyResourceDelta(IResourceDelta delta) {
+		convertToRustFileLocation(delta.getResource()).ifPresent(location -> {
+			switch(delta.getKind()) {
+			case IResourceDelta.ADDED:
+			case IResourceDelta.CHANGED:
+				fileTouched(location);
+				break;
+			case IResourceDelta.REMOVED:
+				fileRemoved(location);
+				break;
+			}
+		});
+		return true;
+	}
+	
+	private static Optional<Location> convertToRustFileLocation(IResource resource) {
+		return Optional.of(resource)
+				.filter(res -> res instanceof IFile)
+				.filter(RustEditorActionContributor::isValidRustPath)
+				.map(ResourceUtils::getResourceLocation);
+	}
+	
+	// TODO: Find a good filter
+	private static boolean isValidRustPath(IResource resource) {
+		String path = resource.getFullPath().toString();
+		return path.contains("src") && path.endsWith(".rs");

```

Comment:
Found out that `IResourceChangeListener`s are only notified about Rust file changes. Obviously, there is some integrated source file concept. Still, I don't know how to apply this concept for the initial indexing run.